Tui 1.0

oliver-sanders opened the pull request:
Remove the "Tui is experimental" notice :tada:

* Move the data update into another process: Closes #3527
  * Tui used to contact workflows in the same process it used to run the application.
  * This meant that the app was held in lockstep with the update. If the update took longer than the refresh time it would fall over.
  * This PR puts the updater into another process to break this lockstep so that the application and updater refresh rates can differ.
* Add multi-workflow capability: Closes #3464
  * Use the `scan` pipe to list workflows in Tui the same way we use it in the CylcUIServer.
  * The workflow refresh rate is different to the scan refresh rate.
  * Each workflow node has one child under it, the "spring" node. When you expand this "spring" node, Tui will connect to the workflow and will replace the "spring" node with the workflow tree once the data arrives.
  * When you collapse the workflow node, Tui will disconnect from the workflow.
  * You can expand multiple workflows.
* Add visual regression testing framework: Closes #3530
  * Add a test fixture which launches Tui in a "headless" mode where it's not rendering to a terminal.
  * The fixture has a method to provide user input allowing you to press keys as if there was a terminal running.
  * The fixture has a method for generating HTML screenshots which are then used for visual regression.
  * I.E. if anything changes in the output HTML, then test will fail.
  * This allows us to test navigation, filtering, help screens and more.
  * Tui coverage is now up to ~80%.
* Add some new functionalities.
  * Log view: support for `cylc cat-log`.
  * Show view: support for `cylc show`.


https://github.com/cylc/cylc-flow/assets/16705946/ff7d491d-9df4-4f66-9cd5-935d8d080c13



In the process I spotted an issue with task filtering which looks like it might be a data store thing but requires more investigation, lower priority as filtering is not an issue for the GUI. Until then the task filtering won't work right: https://github.com/cylc/cylc-flow/issues/5716

TODO:

* [x] Niceify the code.
* [x] Complete restart/reconnect test.
* [x] Do something better with the Tui header (which is now blank).
* [x] Investigate test flakyness (workflow states test quite flaky).

Tui integration test failures result in HTML files like this being written into the test directory in `~cylc-run/`. When tests fail in CI, these HTML files will be bundled into the artefact so can be inspected locally to determine the cause of the failure:

![Screenshot from 2023-09-26 13-17-55](https://github.com/cylc/cylc-flow/assets/16705946/e3ce363e-04b7-44ba-be1f-527c21102b55)

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` (and `conda-environment.yml` if present).
- [x] Tests are included (or explain why tests are not needed).
- [x] `CHANGES.md` entry included if this is a change that can affect users
- [x] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request opened if required at [cylc/cylc-doc/pull/663.](https://github.com/cylc/cylc-doc/pull/663)
- [x] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.

oliver-sanders commented on tests/integration/tui/screenshots/test_navigation.cursor-at-bottom-of-screen.html:
```diff
@@ -0,0 +1,31 @@
+<pre><span style="color:#000000;background:#e5e5e5">                                                                                </span>
```
I was really hoping GitHub would render this HTML which would make it much easier to review screenshot updates in PRs. Any ideas?

oliver-sanders replied:
The best thing I can come up with is to use some Python lib to render them into png derived files.

hjoliver commented:
Nice, looking forward to this!

oliver-sanders commented on cylc/flow/scripts/tui.py:
```diff
@@ -15,34 +15,37 @@
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
-"""cylc tui WORKFLOW
+"""cylc tui [WORKFLOW]
```
If you give a workflow, tui will startup, pre-filtered to just that workflow. Press "E" to unset this filtering an view all workflows.

oliver-sanders commented on cylc/flow/scripts/tui.py:
```diff
@@ -55,57 +58,36 @@
     )
 
-    parser.add_option(
-        '--display',
```
Removed the HTML testing mode from the Tui script as this is now handled by an integration test fixture.

oliver-sanders commented on cylc/flow/tui/app.py:
```diff
@@ -206,16 +246,25 @@ class TuiApp:
 
-    UPDATE_INTERVAL = 1
-    CLIENT_TIMEOUT = 1
+    # the UI update interval
+    # NOTE: this is different from the data update interval
+    UPDATE_INTERVAL = 0.1
```
This means that the UI will look for updates every 0.1 seconds, but the frequency of these updates is determined by the updater which is decoupled and runs to its own clock.

oliver-sanders commented on cylc/flow/tui/data.py:
```diff
@@ -154,23 +155,20 @@ def cli_cmd(*cmd):
-    # for now we will exit tui when the workflow is cleaned
-    # this will change when tui supports multiple workflows
-    cli_cmd('clean', workflow)
-    sys.exit(0)
```
Tui used to exit when you cleaned the workflow, but now it can stay open.

oliver-sanders commented on cylc/flow/tui/updater.py:
```diff
@@ -0,0 +1,311 @@
+
+    """
+
+    # the maximum time to wait for a workflow update
+    CLIENT_TIMEOUT = 2
```
If the workflow takes more than 2 seconds to send its response, then it will time out. Tui will put a message under the workflow saying that it has timed out, but will retry with the next update cycle.

Previously both the application and updater used the same update period of 1 second. If it took longer to get the response and do the required processing, then Tui would crash.

oliver-sanders commented on cylc/flow/tui/updater.py:
```diff
@@ -0,0 +1,311 @@
+    # the interval between workflow listing scans
+    BASE_SCAN_INTERVAL = 20
+
+    # the interval between workflow data updates
+    BASE_UPDATE_INTERVAL = 1
```
This is the desired update interval. If it takes longer than 1 second to update the workflow, no problem, the updates just come through slower, no more crashes.

oliver-sanders commented:
With Tui now better tested, this brings the project coverage up above 90%!

We still can get more of Tui covered now there's a testing framework for it.

dwsutherland commented:
Amazing work 🚀

oliver-sanders commented:
(added some extra tests to fill some coverage holes)

wxtim commented:
Sorry, I think that I found a bug on pressing `p` in the main view I get this error (sorry, my terminal isn't letting me copy/paste it:

![image](https://github.com/cylc/cylc-flow/assets/26465611/14b131e1-6f0f-4922-9e9d-bac0cfcce365)

oliver-sanders commented:
Good spot, the tests cover toggling workflow states using `W` and filling out the form, but not using the shortcuts. Will fix.

wxtim commented:
And from `user>stop-all`

```
Cylc Tui   tasks filtered (F - edit, R - re┌────────────────────────────────────────────────┐
                                           │                                                │
- ~tim                                     │  Error in command cylc stop *                  │
                                           │  ClientError: Failed to find user's private    │
                                           │  key, so cannot connect.                       │
                                           │  Task exception was never retrieved            │
                                           │  future: <Task finished name='Task-124'        │
                                           │  coro=<run() done, defined at /home/tim/tims_  │
                                           │  code/metomi/cylc-flow/cylc/flow/scripts/stop  │
                                      ┌────│  .py:206> exception=ClientError("Failed to     │
                                      │  id│  find user's private key, so cannot            │
                                      │    │  connect.")>                                   │
                                      │  Ac│  Traceback (most recent call last):            │
                                      │  < │    File "/home/tim/tims_code/metomi/cylc-flow  │
                                      │    │  /cylc/flow/network/__init__.py", line 237,    │
                                      │  < │  in _socket_connect                            │
                                      │    │      client_public_key, client_priv_key =      │
                                      │    │  zmq.auth.load_certificate(                    │
                                      │    │                                                │
                                      │    │  ^^^^^^^^^^^^^^^^^^^^^^^^^^                    │
                                      │    │    File "/home/tim/mambaforge/envs/cylc.dev/l  │
                                      │    │  ib/python3.11/site-packages/zmq/auth/certs.p  │
                                      │    │  y", line 105, in load_certificate             │
                                      │    │      raise OSError(f"Invalid certificate       │
                                      │    │  file: {filename}")                            │
                                      │    │  OSError: Invalid certificate file: /home/tim  │
                                      │    │  /cylc-run/flows-hold/run16/.service/client.k  │
                                      │ q t│  ey_secret                                     │
                                      └────│                                                │
                                           │  During handling of the above exception,       │
                                           │  another exception occurred:                   │
                                           │                                                │
                                           │  Traceback (most recent call last):            │
                                           │    File "/home/tim/tims_code/metomi/cylc-flow  │
                                           │  /cylc/flow/scripts/stop.py", line 223, in     │
                                           │                                                │
quit: q  help: h  context: enter  tree: - ←│ q to close                                     │T f s r R  filter workflows: W E p
                                           └────────────────────────────────────────────────┘

```

wxtim commented:
Would `clean-all-stopped` be unreasonable to ask for?

oliver-sanders commented:
> And from user>stop-all

I think that's correct behaviour. The `clean` command failed, so Tui put the stderr in a box for you.

oliver-sanders commented:
> Would clean-all-stopped be unreasonable to ask for?

Not unreasonable, I think that's just `cylc clean '*:stopped'` or since clean auto filters out running workflows, just `cylc clean '*'`.

It might be considered a little dangerous having a button like that kicking about, so we might want to put an "are you sure" check on that which Tui doesn't currently have an interface for.

oliver-sanders commented:
Fixed the "p" shortcut and added it to the tests: https://github.com/cylc/cylc-flow/compare/8e371542150b9792a8593399cefc14f96be5c1df..54b7e609af8676c7fcb3b4e8723a50d92ad9c3a7

oliver-sanders commented:
Fixed an issue with "E" not working properly: https://github.com/cylc/cylc-flow/compare/54b7e609af8676c7fcb3b4e8723a50d92ad9c3a7..6f3f1701c1106ecd779b11b08ad4293186678da0

oliver-sanders commented on cylc/flow/scripts/tui.py:
```diff
@@ -55,57 +57,36 @@
-    workflow_id, *_ = parse_id(
-        workflow_id,
-        constraint='workflows',
+def configure_screenshot(v_term_size):
```
I'll move it in with the test code.

wxtim reviewed (commented):
- [x] Functional testing of PR.
- [x] Read function code.
- [x] Read the testing.

wxtim commented on tests/integration/tui/conftest.py:
```diff
@@ -0,0 +1,266 @@
+SCREENSHOT_DIR = Path(__file__).parent / 'screenshots'
+
+
+def format_test_failure(expected, got, description):
```
Can we have a docstring? Why are we turning the diff into an HTML doc?

wxtim commented on tests/integration/tui/screenshots/test_auto_expansion.later-time.html:
```diff
@@ -0,0 +1,21 @@
+<pre><span style="color:#ff0000;background:#e5e5e5;font-weight:bold">C</span><span style="color:#ffff00;background:#e5e5e5;font-weight:bold">y</span><span style="color:#00ff00;background:#e5e5e5;font-weight:bold">l</span><span style="color:#5c5cff;background:#e5e5e5;font-weight:bold">c</span><span style="color:#000000;background:#e5e5e5;font-weight:bold"> Tui</span><span style="color:#7f7f7f;back ...
```
Are you OK with this offest?

![Untitled](https://github.com/cylc/cylc-flow/assets/26465611/8f9d9f76-4f33-455d-a16c-163663fbf522)


(it appears in lots of your images)

oliver-sanders replied:
The reason for this is the way your browser handles the queued modifier.

In XTerm, the modifier applies to the task icon with a character width of 1. In FireFox the modifier comes before the task icon with a character width of 2.

wxtim commented on tests/integration/tui/test_app.py:
```diff
@@ -0,0 +1,378 @@
+            'the context menu should have closed',
+        )
+
+        # "ctrl d" should exit Tui
```
Ought you test `ctrl c` and `ctrl z` cases too?

oliver-sanders replied:
`ctrl z` can't be reasonably tested (you need a genuine interactive terminal).

`ctrl c` doesn't yield the exception.

oliver-sanders commented:
Training day came around again, added support for `cat-log` and `show`.

@wxtim, could you check the last three commits.

oliver-sanders commented on cylc/flow/data_store_mgr.py:
```diff
@@ -1380,7 +1380,7 @@ def insert_job(self, name, cycle_point, status, job_conf):
             cycle_point=tproxy.cycle_point,
             execution_time_limit=job_conf.get('execution_time_limit'),
-            platform=job_conf.get('platform')['name'],
+            platform=job_conf['platform']['name'],
```
Unrelated, spotted during testing.

The `.get` here wasn't protecting against anything as the default return value of `None` would cause the `['name']` bit to error anyway.

wxtim reviewed (approved):
This is great. Please stop with the mission creep so that we can merge. Pretty Please.
